components/compliance-service/ingest/ingestic/ingestic.go

```diff
@@ -416,6 +424,10 @@ func (backend *ESClient) UpdateSummaryProjectsTags(ctx context.Context, projectT
 						}
 					}
 					if (condition.roles.length > 0) {
```

Comment:
Since each attribute has an identical code block pattern--and indeed identical code across `UpdateReportProjectsTags`, `UpdateSummaryProjectsTags`, and `UpdateNodeProjectTags`--you could reduce a lot of duplication by introducing two simple code generation functions, so that `script` in this function could be reduced to this, making it easier to maintain:
```
script := `
		ArrayList matchingProjects = new ArrayList();
		for (def project : params.projects) {
			for (def rule : project.rules) {
				def match = rule.conditions.length != 0;
				for (def condition : rule.conditions) {` +
					genSingleLoop("chefServer", "source_fqdn") + 
					genSingleLoop("environment", "environment") + 
					genSingleLoop("organization", "organization_name") + 
					genSingleLoop("policyName", "policy_name") + 
					genSingleLoop("policyGroup", "policy_group") + 
					genDoubleLoop("roles", "roles", "Role") + 
					genDoubleLoop("chefTags", "chef_tags", "ChefTag") + `
				}
				if ( match ) {
					matchingProjects.add(project.name);
					break;
				}
			}
		}
		ctx._source.projects = matchingProjects.toArray();
`
```

Reply:
How do you create a function in the painless script?